In [ ]:
import importlib
import subprocess
import sys
import gc

def check_and_install_package(package_name, version=None):
    try:
        importlib.import_module(package_name)
        print(f"\n{package_name} is already installed.")
    except ImportError:
        print(f"\n{package_name} is NOT installed. Installing now...")
        if version:
            subprocess.check_call([sys.executable, "-m", "pip", "install", f"{package_name}=={version}"])
        else:
            subprocess.check_call([sys.executable, "-m", "pip", "install", package_name])
        print(f"{package_name} installation completed.")

# List of packages to check along with specific versions if necessary
packages = [
    {"name": "tqdm", "version": None},
    {"name": "pyspark", "version": "3.5.2"},
    {"name": "gdown", "version": None},
    {"name": "numpy", "version": "1.23.5"}
]

# Checking and installing the packages
for package in packages:
    check_and_install_package(package["name"], package["version"])


tqdm is already installed.

pyspark is already installed.

gdown is already installed.

numpy is already installed.


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
from pyspark.sql import SparkSession

spark = SparkSession.builder \
    .appName("GBTModel") \
    .config("spark.driver.memory", "16g") \
    .config("spark.executor.memory", "16g") \
    .config("spark.driver.maxResultSize", "8g") \
    .config("spark.executor.memoryOverhead", "12g") \
    .config("spark.executor.cores", "5") \
    .config("spark.serializer", "org.apache.spark.serializer.KryoSerializer") \
    .config("spark.kryoserializer.buffer.max", "2047m") \
    .config("spark.dynamicAllocation.enabled", "true") \
    .config("spark.hadoop.fs.file.impl", "org.apache.hadoop.fs.LocalFileSystem") \
    .config("spark.executor.extraJavaOptions", "-XX:+UseG1GC -XX:InitiatingHeapOccupancyPercent=35 -XX:ConcGCThreads=4 -XX:ParallelGCThreads=4") \
    .getOrCreate()

# Verifying Spark session creation
print(f"Spark session started with version: {spark.version}")


Spark session started with version: 3.5.3


# **Predictions BEFORE Feature Engineering**

In [ ]:
# loading the df

!cp '/content/drive/MyDrive/Big Data Analytics - Project/Datasets/Processed_DF.parquet' /content/

output_path = '/content/Processed_DF.parquet'
df = spark.read.parquet(output_path)
print("The Processed DataFrame has been loaded successfully.")


The Processed DataFrame has been loaded successfully.


In [ ]:
df = df.repartition(100)  # Repartitioning into 100 partitions for parallelism

In [ ]:
# Printing the shape of the DataFrame
total_rows = df.count()
total_columns = len(df.columns)

print(f"The shape of the loaded DataFrame is: ({total_rows}, {total_columns})")

The shape of the loaded DataFrame is: (3000040, 42)


In [ ]:
from functools import reduce
from pyspark.sql.functions import col

# Checking for rows with at least one null value in any column
rows_with_null = df.filter(
    reduce(lambda x, y: x | y, (col(c).isNull() for c in df.columns))
).count()

print(f"Number of rows with at least one null value: {rows_with_null}")

Checking for Null values -> Number of rows with at least one null value: 0


### **Handling Categorical Coloumns**

In [ ]:
df=df.drop('description','major_options','mileage')
# Keeping the columns ['exterior_color','dealer_zip','interior_color']

In [ ]:
# Counting unique values in 'exterior_color' and 'interior_color' columns
exterior_colors_count = df.select('exterior_color').distinct().count()
interior_colors_count = df.select('interior_color').distinct().count()

print(f"Unique exterior colors: {exterior_colors_count}")
print(f"Unique interior colors: {interior_colors_count}")

Unique exterior colors: 23036
Unique interior colors: 38528


In [ ]:
from pyspark.sql import functions as F
from pyspark.sql.types import ArrayType, StringType

# Listing top colors for both exterior and interior
colors = ['White', 'Black', 'Gray', 'Silver', 'Red', 'Blue', 'Brown', 'Green', 'Beige', 'Orange', 'Gold', 'Yellow', 'Purple']

# Creating a UDF to find colors in the color columns
@F.udf(returnType=ArrayType(StringType()))
def find_colors(color_string):
    if color_string is None or color_string.strip() == "":
        return ["Other"]  # Handle empty or null values
    found_colors = [c for c in colors if c.lower() in color_string.lower()]
    return found_colors if found_colors else ["Other"]  # Label non-matching colors as "Other"

# Applying the UDF to both the exterior and interior color columns
df = df.withColumn("exterior_color_array", find_colors("exterior_color"))
df = df.withColumn("interior_color_array", find_colors("interior_color"))

# Creating a column with the count of colors found for both exterior and interior
df = df.withColumn("exterior_color_count", F.size("exterior_color_array"))
df = df.withColumn("interior_color_count", F.size("interior_color_array"))

# Joining the color arrays into string columns
df = df.withColumn("exterior_color", F.array_join("exterior_color_array", ", "))
df = df.withColumn("interior_color", F.array_join("interior_color_array", ", "))

# Labeling mixed colors for both exterior and interior colors
df = df.withColumn(
    "exterior_color",
    F.when(F.col("exterior_color_count") > 1, "Mixed Colors")
     .otherwise(F.col("exterior_color")))

df = df.withColumn(
    "interior_color",
    F.when(F.col("interior_color_count") > 1, "Mixed Colors")
     .otherwise(F.col("interior_color")))

# Dropping temporary columns
df = df.drop("exterior_color_array", "exterior_color_count", "interior_color_array", "interior_color_count")


In [ ]:
# Counting the occurrences of each exterior and interior color and calculating percentages
exterior_color_counts = df.groupBy("exterior_color").count().withColumn(
    "percentage", F.round((F.col("count") / df.count()) * 100, 2))

interior_color_counts = df.groupBy("interior_color").count().withColumn(
    "percentage", F.round((F.col("count") / df.count()) * 100, 2))

# Showing the results
print("Exterior Color Distribution:")
exterior_color_counts.orderBy(F.desc("count")).show(truncate=False)

Exterior Color Distribution:
+--------------+------+----------+
|exterior_color|count |percentage|
+--------------+------+----------+
|White         |675979|22.53     |
|Black         |580148|19.34     |
|Other         |543638|18.12     |
|Silver        |384540|12.82     |
|Blue          |253263|8.44      |
|Red           |242331|8.08      |
|Gray          |231172|7.71      |
|Green         |23026 |0.77      |
|Mixed Colors  |19728 |0.66      |
|Brown         |12905 |0.43      |
|Orange        |11638 |0.39      |
|Gold          |10544 |0.35      |
|Beige         |5065  |0.17      |
|Yellow        |4855  |0.16      |
|Purple        |1208  |0.04      |
+--------------+------+----------+



In [ ]:
print("Interior Color Distribution:")
interior_color_counts.orderBy(F.desc("count")).show(truncate=False)

Interior Color Distribution:
+--------------+-------+----------+
|interior_color|count  |percentage|
+--------------+-------+----------+
|Black         |1624033|54.13     |
|Other         |577578 |19.25     |
|Gray          |383966 |12.8      |
|Mixed Colors  |171212 |5.71      |
|White         |91545  |3.05      |
|Brown         |65943  |2.2       |
|Red           |34117  |1.14      |
|Silver        |24124  |0.8       |
|Blue          |22828  |0.76      |
|Green         |2048   |0.07      |
|Gold          |1193   |0.04      |
|Orange        |1133   |0.04      |
|Yellow        |134    |0.0       |
|Purple        |121    |0.0       |
|Beige         |65     |0.0       |
+--------------+-------+----------+



In [ ]:
print(f"Final processed DataFrame used for the model has {df.count()} rows and {len(df.columns)} columns.")

Final processed DataFrame used for the model has 3000040 rows and 39 columns.


In [ ]:
# Calculating the average price
avg_price = df.agg({"price": "avg"}).collect()[0][0]
print(f"Average price of a car: {round(avg_price)}")

Average price of a car: 29933


In [ ]:
import pandas as pd
from IPython.display import display
import pyspark.sql.functions as F

# Converting the Spark DataFrame to a Pandas DataFrame and displaying 5 random rows with all columns
pd.set_option('display.max_columns', None)
pandas_df = df.orderBy(F.rand()).limit(5).toPandas()
display(pandas_df)


fuel_type        body_type          city  city_fuel_economy  daysonmarket  \
0  Gasoline            Wagon      Gulfport          17.000000            42   
1  Gasoline          Unknown         Tampa          22.690001             7   
2  Gasoline  SUV / Crossover        Orange          23.000000            11   
3    Diesel              Van  Watkinsville          22.690001            76   
4  Gasoline            Sedan          Avon          28.000000            29   

  dealer_zip  engine_displacement      engine_type exterior_color  \
0      39507               3500.0               V6          Other   
1      33619               2500.0  Gasoline engine          White   
2      92867               2500.0               I4           Gray   
3      30677               3000.0               V6          Other   
4      46123               2000.0               I4           Blue   

   franchise_dealer  fuel_tank_volume  height  highway_fuel_economy  \
0             False             18.00   58.30             24.000000   
1              True             18.09   65.87             29.469999   
2              True             15.90   67.10             30.000000   
3              True             24.50   96.30             29.469999   
4              True             12.40   57.00             37.000000   

   horsepower interior_color  is_new   latitude  length listed_date  \
0       250.0          Other   False  30.391500   197.7  2020-07-31   
1       241.0          Black    True  27.948000   189.8  2020-09-03   
2       176.0          Black   False  33.810799   183.5  2020-08-31   
3       188.0          Black    True  33.914902   233.5  2020-06-26   
4       149.0           Blue    True  39.763199   182.7  2020-08-13   

  listing_color   longitude      make_name  maximum_seating model_name  \
0       UNKNOWN  -89.059097          Dodge              5.0     Magnum   
1         WHITE  -82.343300          Mazda              5.0      CX-30   
2          GRAY -117.835999         Toyota              5.0       RAV4   
3       UNKNOWN  -83.473198  Mercedes-Benz              3.0   Sprinter   
4          BLUE  -86.364502         Nissan              5.0     Sentra   

     price  savings_amount  seller_rating                  sp_name  torque  \
0   3000.0               0       5.000000              Mad Kustomz  250.00   
1  27795.0               0       5.000000  Ferman Mazda of Brandon  265.22   
2  21899.0            1893       4.222222         Toyota of Orange  172.00   
3  47985.0               0       4.476191  Mercedes-Benz of Athens  325.00   
4  21003.0               0       3.734940    Andy Mohr Avon Nissan  146.00   

  transmission                transmission_display wheel_system_display  \
0            A                   5-Speed Automatic     Rear-Wheel Drive   
1            A                           Automatic              Unknown   
2            A                           Automatic    Front-Wheel Drive   
3            A                   7-Speed Automatic                  4X2   
4          CVT  Continuously Variable Transmission    Front-Wheel Drive   

   wheelbase  width  year  combined_fuel_economy  legroom  log_mileage  \
0      120.0   74.1  2007                  20.50    82.00        12.62   
1      111.0   77.2  2021                  26.08    80.16         8.91   
2      104.7   72.6  2017                  26.50    79.80        10.26   
3      144.3   92.3  2020                  26.08    80.16         2.30   
4      106.8   71.5  2020                  32.50    81.40         8.91   

   major_options_count  
0                    0  
1                    9  
2                    4  
3                    3  
4                    5



---



# **GBT Regressor**

**DEMO Code for checking if libraries are properly installed**

In [ ]:
from pyspark.ml.regression import GBTRegressor
from pyspark.ml.linalg import Vectors
from pyspark.sql import Row
from pyspark.ml.evaluation import RegressionEvaluator

# Sample data creation: creating a simple DataFrame with features and a label for regression
data1 = [
    Row(label=1.0, features=Vectors.dense([1.0, 2.0, 3.0])),
    Row(label=2.0, features=Vectors.dense([4.0, 5.0, 6.0])),
    Row(label=3.0, features=Vectors.dense([7.0, 8.0, 9.0])),
    Row(label=4.0, features=Vectors.dense([10.0, 11.0, 12.0]))
]

df1 = spark.createDataFrame(data1)

# Splitting the data into train and test sets
train_df1, test_df1 = df1.randomSplit([0.8, 0.2], seed=42)

# Defining GBTRegressor
gbt_regressor = GBTRegressor(
    featuresCol="features",
    labelCol="label",
    maxIter=10,
    maxDepth=3
)

# Training the model
model = gbt_regressor.fit(train_df1)

# Making predictions
predictions = model.transform(test_df1)

# Showing predictions
predictions.select("features", "label", "prediction").show()

# Evaluating the model using RegressionEvaluator
evaluator = RegressionEvaluator(labelCol="label", predictionCol="prediction", metricName="rmse")
rmse = evaluator.evaluate(predictions)

print(f"\nRoot Mean Squared Error (RMSE): {rmse:.4f}")


+-------------+-----+----------+
|     features|label|prediction|
+-------------+-----+----------+
|[1.0,2.0,3.0]|  1.0|       2.0|
+-------------+-----+----------+


Root Mean Squared Error (RMSE): 1.0000




---



### **On 20% of the Dataset**

In [ ]:
import warnings
from tqdm import tqdm
from pyspark.ml.feature import StringIndexer, VectorAssembler, StandardScaler, OneHotEncoder
from pyspark.ml import Pipeline
from pyspark.sql.functions import mean as sql_mean
import pyspark.sql.functions as F

# Ignore warnings
warnings.filterwarnings('ignore')

print("Processing the data...")
with tqdm(total=5, desc="Progress") as pbar:


    df_sample = df.sample(fraction=0.2, seed=42)  # Randomly sample 600k records of the data
    pbar.update(1)

    # Handling categorical columns
    cat_columns = [field for (field, dtype) in df_sample.dtypes if dtype == "string"]
    stages = []
    for col_name in cat_columns:
        indexer = StringIndexer(inputCol=col_name, outputCol=f"{col_name}_indexed", handleInvalid="keep")
        encoder = OneHotEncoder(inputCol=f"{col_name}_indexed", outputCol=f"{col_name}_encoded")
        stages += [indexer, encoder]
    pbar.update(1)

    # Assembling features
    num_columns = [col for col in df_sample.columns if col != 'price' and col not in cat_columns]
    encoded_columns = [f"{col}_encoded" for col in cat_columns]
    feature_columns = num_columns + encoded_columns
    assembler = VectorAssembler(inputCols=feature_columns, outputCol="features")
    stages += [assembler]
    pbar.update(1)

    # Adding scaling to the pipeline
    scaler = StandardScaler(inputCol="features", outputCol="scaled_features", withMean=True, withStd=True)
    stages += [scaler]

    # Creating and applying the pipeline
    pipeline = Pipeline(stages=stages)
    pipeline_model = pipeline.fit(df_sample)
    df_sample = pipeline_model.transform(df_sample)
    pbar.update(1)

    # Splitting the data
    train_df, test_df = df_sample.randomSplit([0.8, 0.2], seed=42)
    pbar.update(1)

print("\n\nData preprocessing and splitting completed!")


Processing the data...


Progress: 100%|██████████| 5/5 [00:45<00:00,  9.05s/it]



Data preprocessing and splitting completed!


In [ ]:
from pyspark.ml.regression import GBTRegressor
from pyspark.ml.evaluation import RegressionEvaluator
import time

# Starting to track overall run time
start_time = time.time()

# Model training
print("Training GBTRegressor model...")

# Using GBTRegressor with tuned parameters
gbt_regressor = GBTRegressor(
    featuresCol="scaled_features",
    labelCol="price",
    maxIter=100,
    maxDepth=5,
    seed=42,
)

# Training the model
model = gbt_regressor.fit(train_df)

# Making predictions
print("Making predictions...")
predictions = model.transform(test_df)

# Evaluating the model
print("Evaluating the model...")
evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="r2")
r2 = evaluator.evaluate(predictions)

print(f"\nTrain size: {train_df.count():,} samples")
print(f"Test size: {test_df.count():,} samples")

# Multiplying R-Squared by 100 and rounding it off
print(f"\n\nR-Squared Score (Accuracy): {round(r2 * 100)}%\n")

# Calculating total runtime
end_time = time.time()
total_runtime = (end_time - start_time) / 60
print(f"\nOverall runtime: {round(total_runtime)} minutes.")


Training GBTRegressor model...
Making predictions...
Evaluating the model...

Train size: 480,538 samples
Test size: 120,299 samples


R-Squared Score (Accuracy): 83.21%


Overall runtime: 222 minutes.


In [ ]:
# Calculate additional metrics
mae_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="mae")
mae = mae_evaluator.evaluate(predictions)

mse_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="mse")
mse = mse_evaluator.evaluate(predictions)

rmse_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="rmse")
rmse = rmse_evaluator.evaluate(predictions)

print("Additional Metrics:")
print(f"Mean Absolute Error: {round(mae)}")
print(f"Mean Squared Error: {round(mse)}")
print(f"Root Mean Squared Error: {round(rmse)}")


Additional Metrics:
Mean Absolute Error: 3578
Mean Squared Error: 63098842
Root Mean Squared Error: 7943


The Mean Absolute Error (`MAE`) of **\$3578**  suggests that, on average, the predicted car prices deviate from the actual prices by this amount. Given that the Mean car price is **\$29,933** , this error represents about `12.0% of the mean price`, which suggests that the model performs well.

In [ ]:
# Feature Importance
feature_importance = model.featureImportances
features_list = pipeline_model.stages[-2].getInputCols()  # Getting feature names from VectorAssembler

# Sorting features by their importance in descending order
sorted_features = [feature for feature, importance in sorted(zip(features_list, feature_importance), key=lambda x: x[1], reverse=True)]

# Printing ranked features from highest to lowest
print("Top 10 Features Ranked by Importance (Highest to Lowest):")
for rank, feature in enumerate(sorted_features[:10], 1):  # Limiting to top 10 features
    print(f"{rank}. {feature}")

Top 10 Features Ranked by Importance (Highest to Lowest):
1. horsepower
2. log_mileage
3. highway_fuel_economy
4. year
5. engine_displacement
6. fuel_tank_volume
7. major_options_count
8. torque
9. width
10. maximum_seating


In [ ]:
import pyspark

# Getting the list of all variables in the current notebook environment
for var_name in dir():
    # Checking if the variable is a Spark DataFrame
    if isinstance(eval(var_name), pyspark.sql.dataframe.DataFrame):
        # Deleting the variables
        del globals()[var_name]

# Performing garbage collection
import gc
_ = gc.collect()
print("All DataFrames have been deleted.")


spark.catalog.clearCache()
spark.stop()
print("Spark Stopped !")

All DataFrames have been deleted.
Spark Stopped !




---



# **Predictions AFTER Feature Engineering**

In [ ]:
!cp '/content/drive/MyDrive/Big Data Analytics - Project/Datasets/Feature_Engineered_DF.parquet' /content/

output_path = '/content/Feature_Engineered_DF.parquet'
df = spark.read.parquet(output_path)
print("The Feature Engineered DataFrame has been loaded successfully.")


The Feature Engineered DataFrame has been loaded successfully.


In [ ]:
# Printing the shape of the DataFrame
total_rows = df.count()
total_columns = len(df.columns)

print(f"The shape of the loaded DataFrame is: ({total_rows}, {total_columns})")

The shape of the loaded DataFrame is: (3000040, 47)


In [ ]:
# Calculating the average price
avg_price = df.agg({"price": "avg"}).collect()[0][0]
print(f"Average price of a car: {round(avg_price)}")

Average price of a car: 29933


In [ ]:
import pandas as pd
from IPython.display import display
import pyspark.sql.functions as F

# Converting the Spark DataFrame to a Pandas DataFrame and displaying 5 random rows with all columns
pd.set_option('display.max_columns', None)
pandas_df = df.orderBy(F.rand()).limit(5).toPandas()
display(pandas_df)


fuel_type        body_type          city  city_fuel_economy  days_in_market  \
0  Gasoline            Sedan       Lilburn               18.0              48   
1  Gasoline      Convertible     Dade City               22.0              13   
2  Gasoline  SUV / Crossover       Fontana               25.0              12   
3  Gasoline  SUV / Crossover  Jacksonville               17.0              64   
4  Gasoline  SUV / Crossover     Flagstaff               18.0               9   

  dealer_zip  engine_displacement engine_type exterior_color  \
0      30047               3800.0          V6          Black   
1      33525               2000.0          I4          Black   
2      92336               2500.0          I4           Gray   
3      32244               3000.0          V6           Blue   
4      86004               3600.0          V6          White   

   franchise_dealer  fuel_tank_volume  height  highway_fuel_economy  \
0             False         20.299999    58.3                  29.0   
1              True         17.400000    55.5                  29.0   
2              True         14.800000    66.2                  31.0   
3              True         22.500000    68.5                  21.0   
4              True         24.600000    69.3                  25.0   

   horsepower interior_color  is_new   latitude  length listing_color  \
0       311.0   Mixed Colors   False  33.899899   196.5         BLACK   
1       241.0   Mixed Colors   False  28.332600   184.5         BLACK   
2       187.0          Black    True  34.133499   179.1          GRAY   
3       335.0          Black    True  30.215099   199.3          BLUE   
4       295.0          Black    True  35.223202   189.8         WHITE   

    longitude      make_name  maximum_seating      model_name    price  \
0  -84.128998        Hyundai              5.0         Genesis  21089.0   
1  -82.186600  Mercedes-Benz              4.0         C-Class  39544.0   
2 -117.438004          Mazda              5.0            CX-5  26730.0   
3  -81.738403           Audi              7.0              Q7  78460.0   
4 -111.568001           Jeep              5.0  Grand Cherokee  37023.0   

   savings_amount  seller_rating                         sp_name  torque  \
0               0       3.914286               Atlanta Used Cars   293.0   
1            2668       4.909091  Jim Browne Chrysler Jeep Dodge   273.0   
2               0       3.628572                   Fontana Mazda   186.0   
3               0       4.105263                Audi Orange Park   369.0   
4               0       3.400000  Planet Chrysler Dodge Jeep Ram   260.0   

  transmission transmission_display wheel_system_display  wheelbase  width  \
0            A            Automatic     Rear-Wheel Drive      118.5   74.4   
1            A    9-Speed Automatic      All-Wheel Drive      111.8   79.3   
2            A    6-Speed Automatic    Front-Wheel Drive      106.2   83.3   
3            A    8-Speed Automatic      All-Wheel Drive      117.9   87.1   
4            A    8-Speed Automatic     Four-Wheel Drive      114.7   84.8   

   manufactured_year  combined_fuel_economy  legroom  log_mileage  \
0               2016                   23.5     80.7        10.84   
1               2017                   25.5     74.0        10.31   
2               2020                   28.0     80.6         2.30   
3               2020                   19.0     80.5         2.40   
4               2020                   21.5     78.9         0.00   

   major_options_count  hp_x_engine_disp  hp_x_torque  listed_day  \
0                    7              0.47      0.18430          24   
1                    5              0.05     -0.00537          28   
2                    4              0.23      0.50223          30   
3                   10              0.04      0.94909           8   
4                    8              0.27     -0.02589           2   

   listed_month  listed_year  age  resale_value_score  maintena

In [ ]:
import warnings
from tqdm import tqdm
from pyspark.ml.feature import StringIndexer, VectorAssembler, StandardScaler, OneHotEncoder
from pyspark.ml import Pipeline
from pyspark.sql.functions import mean as sql_mean
import pyspark.sql.functions as F

# Ignore warnings
warnings.filterwarnings('ignore')

print("Processing the data...")
with tqdm(total=5, desc="Progress") as pbar:

    df_sample = df.sample(fraction=0.2, seed=42)  # Randomly sample 600k records of the data
    pbar.update(1)

    # Handling categorical columns
    cat_columns = [field for (field, dtype) in df_sample.dtypes if dtype == "string"]
    stages = []
    for col_name in cat_columns:
        indexer = StringIndexer(inputCol=col_name, outputCol=f"{col_name}_indexed", handleInvalid="keep")
        encoder = OneHotEncoder(inputCol=f"{col_name}_indexed", outputCol=f"{col_name}_encoded")
        stages += [indexer, encoder]
    pbar.update(1)

    # Assembling features
    num_columns = [col for col in df_sample.columns if col != 'price' and col not in cat_columns]
    encoded_columns = [f"{col}_encoded" for col in cat_columns]
    feature_columns = num_columns + encoded_columns
    assembler = VectorAssembler(inputCols=feature_columns, outputCol="features")
    stages += [assembler]
    pbar.update(1)

    # Adding scaling to the pipeline
    scaler = StandardScaler(inputCol="features", outputCol="scaled_features", withMean=True, withStd=True)
    stages += [scaler]

    # Creating and applying the pipeline
    pipeline = Pipeline(stages=stages)
    pipeline_model = pipeline.fit(df_sample)
    df_sample = pipeline_model.transform(df_sample)
    pbar.update(1)

    # Splitting the data
    train_df, test_df = df_sample.randomSplit([0.8, 0.2], seed=42)
    pbar.update(1)

print("\n\nData preprocessing and splitting completed!")


Processing the data...


Progress: 100%|██████████| 5/5 [00:19<00:00,  3.88s/it]



Data preprocessing and splitting completed!


In [ ]:
# Print the shape of the DataFrame
print(f"The shape of the train_df is: ({train_df.count()}, {len(train_df.columns)})")

The shape of the train_df is: (480411, 77)


In [ ]:
from pyspark.ml.regression import GBTRegressor
from pyspark.ml.evaluation import RegressionEvaluator
import time

# Starting to track overall run time
start_time = time.time()

# Model training
print("Training GBTRegressor model...")

# Using GBTRegressor with tuned parameters
gbt_regressor = GBTRegressor(
    featuresCol="scaled_features",
    labelCol="price",
    maxIter=100,
    maxDepth=5,
    seed=42,
)

# Training the model
model = gbt_regressor.fit(train_df)

# Making predictions
print("Making predictions...")
predictions = model.transform(test_df)

# Evaluating the model
print("Evaluating the model...")
evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="r2")
r2 = evaluator.evaluate(predictions)

print(f"\nTrain size: {train_df.count():,} samples")
print(f"Test size: {test_df.count():,} samples")

# Multiplying R-Squared by 100 and rounding it off
print(f"\n\nR-Squared Score (Accuracy): {round(r2 * 100)}%\n")

# Calculating total runtime
end_time = time.time()
total_runtime = (end_time - start_time) / 60
print(f"\nOverall runtime: {round(total_runtime)} minutes.")


Training GBTRegressor model...
Making predictions...
Evaluating the model...

Train size: 480,411 samples
Test size: 120,366 samples


R-Squared Score (Accuracy): 88.57%


Overall runtime: 678 minutes.


In [ ]:
new_model_path = "/content/drive/MyDrive/Big Data Analytics - Project/models/GBT_model_FE"

# Saving the trained model to the new path
model.save(new_model_path)

print(f"Model saved successfully at {new_model_path}")

Model saved successfully at /content/drive/MyDrive/Big Data Analytics - Project/models/GBT_model_FE


In [ ]:
# Calculate additional metrics
mae_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="mae")
mae = mae_evaluator.evaluate(predictions)

mse_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="mse")
mse = mse_evaluator.evaluate(predictions)

rmse_evaluator = RegressionEvaluator(labelCol="price", predictionCol="prediction", metricName="rmse")
rmse = rmse_evaluator.evaluate(predictions)

print("Additional Metrics:")
print(f"Mean Absolute Error: {round(mae)}")
print(f"Mean Squared Error: {round(mse)}")
print(f"Root Mean Squared Error: {round(rmse)}")

Additional Metrics:
Mean Absolute Error: 3649
Mean Squared Error: 37663588
Root Mean Squared Error: 6137


In **GBT Regressor**, the model saw an improvement in **R² accuracy by 5.36%**, while the Mean Absolute Error **(MAE) slightly increased by 0.22%**, and the Root Mean Square Error **(RMSE) was reduced by approximately 6%**, showing modest gains in predictive performance.

In [ ]:
# Feature Importance (for GBT)
feature_importance = model.featureImportances
features_list = pipeline_model.stages[-2].getInputCols()  # Getting feature names from VectorAssembler

# Sorting features by their importance in descending order
sorted_features = [feature for feature, importance in sorted(zip(features_list, feature_importance), key=lambda x: x[1], reverse=True)]

# Printing ranked features from highest to lowest
print("Top 10 Features Ranked by Importance (Highest to Lowest):")
for rank, feature in enumerate(sorted_features[:10], 1):  # Limiting to top 10 features
    print(f"{rank}. {feature}")

Top 10 Features Ranked by Importance (Highest to Lowest):
1. maintenance_cost
2. log_mileage
3. horsepower
4. engine_displacement
5. torque
6. hp_x_engine_disp
7. maximum_seating
8. major_options_count
9. fuel_tank_volume
10. city_fuel_economy
